packages/sign-client/test/concurrency/concurrency.spec.ts

```diff
@@ -9,57 +9,102 @@ import {
     const relayUrl = process.env.RELAY_URL || TEST_SIGN_CLIENT_OPTIONS.relayUrl;
     const pairings: any[] = [];
 
-    // init clients and pair
-    while (pairings.length < connections) {
-      const clients: Clients = await initTwoClients({ relayUrl });
-
-      expect(clients.A instanceof SignClient).to.eql(true);
-      expect(clients.B instanceof SignClient).to.eql(true);
-
-      const { sessionA } = await testConnectMethod(clients);
-      pairings.push({ clients, sessionA });
-
-      await new Promise<void>(resolve =>
-        setTimeout(() => {
-          resolve();
-        }, 50),
-      );
-    }
-
-    // test communication between clients
-    for (const data of pairings) {
+    const processMessages = async (data: any) => {
       const { clients, sessionA } = data;
-
       const eventPayload: any = {
         topic: sessionA.topic,
         ...TEST_EMIT_PARAMS,
       };
 
       await new Promise<void>(async (resolve, reject) => {
         try {
-          clients.B.on("session_event", (event: any) => {
-            expect(TEST_EMIT_PARAMS).to.eql(event.params);
-            expect(eventPayload.topic).to.eql(event.topic);
-            resolve();
-          });
+          const messagesReceived: any = [];
+          const clientsArr = [clients.A, clients.B];
+
+          for await (const client of [clients.A, clients.B]) {
+            client.on("session_ping", (event: any) => {
+              expect(sessionA.topic).to.eql(event.topic);
+              messagesReceived.push(event);
+              validate();
+            });
+
+            client.on("session_event", (event: any) => {
+              expect(TEST_EMIT_PARAMS).to.eql(event.params);
+              expect(eventPayload.topic).to.eql(event.topic);
+              messagesReceived.push(event);
+              validate();
+            });
+
+            client.on("session_update", (event: any) => {
+              expect(client.session.get(sessionA.topic).namespaces).to.eql(namespacesAfter);
+              messagesReceived.push(event);
+              validate();
+            });
+          }
+
+          const namespacesBefore = sessionA.namespaces;
+          const namespacesAfter = {
+            ...namespacesBefore,
+            eip9001: {
+              accounts: ["eip9001:1:0x000000000000000000000000000000000000dead"],
+              methods: ["eth_sendTransaction"],
+              events: ["accountsChanged"],
+            },
+          };
+
+          const emit = async (client: any) => {
+            await client.emit(eventPayload);
+          };
 
-          clients.A.emit(eventPayload);
+          const validate = () => {
+            if (messagesReceived.length >= messagesToBeExchanged) {
+              resolve();
+            }
+          };
+
+          for (const i of Array.from(Array(messagesToBeExchanged).keys())) {
+            const client: any = Math.floor(Math.random() * clientsArr.length);
```

Comment:
i don't get this?

Reply:
it chooses client randomly to send a message from
so it tests messages between the two clients in both directions
A -> B
B -> A